### PR TITLE
Multi-target fix

### PR DESCRIPTION
**Pull request by Dreamlone**
There was implemented new case with river levels forecasting. Add new case with multi-target regression task with chain composing. The innability of composer to process multitarget regression task was fixed. 

Also, there were investigated several papers aboul multi-tartget metric estimation. Good one can be accesed via [link](https://lirias.kuleuven.be/retrieve/584987/).

### CODE REVIEW

File: cases/multi_target_levels_forecasting.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    forecast = np.ravel(np.array(predicted_output.predict))
+
+    mse_value = mean_squared_error(np.ravel(test_data.target), forecast, squared=False)
+    mae_value = mean_absolute_error(np.ravel(test_data.target), forecast)
+
+    print(f'MAE - {mae_value:.2f}')
+    print(f'RMSE - {mse_value:.2f}\n')
+
+    if vis:
+        plot_predictions(predicted_output, test_data)
```
Comment (nicl-nno) [inline cases/multi_target_levels_forecasting.py:128]: Тут нужен более лаконичный пример с использованием API.
Если текущее не справляется - можно сделать это отдельно в рамках https://github.com/nccr-itmo/FEDOT/issues/306

File: cases/river_levels_prediction/river_level_case_composer.py [hunk 2]
```diff
@@ -28,10 +28,10 @@ def get_chain_info(chain):
     :return depth: depth of the chain
     """
 
-    print('\nObtained chain for current iteration')
+    print('\nObtained chain:')
     obtained_operations = []
     for node in chain.nodes:
-        print(str(node))
+        print(f'{node.operation.operation_type} - {node.custom_params}')
```
Comment (nicl-nno) [inline cases/river_levels_prediction/river_level_case_composer.py:34]: А нужны ли вообще эти принты в рамках кейса? Если это удобный способ печатать структуру chain- можно создать соответствующий метод (или изменить _str__).

File: fedot/core/composer/metrics.py [hunk 3]
```diff
@@ -44,11 +44,22 @@ def get_value(cls, chain: Chain, reference_data: InputData) -> float:
...
                 results.predict = forecast_values
                 metric = cls.metric(reference_data, results)
+            elif is_regression and is_multi_target:
+                # Predictions convert into uni-variate array
+                forecast_values = np.ravel(np.array(results.predict))
+                results.predict = forecast_values
+                # Target convert into uni-variate array
+                target_values = np.ravel(np.array(reference_data.target))
+                reference_data.target = target_values
+                metric = cls.metric(reference_data, results)
```
Comment (nicl-nno) [inline fedot/core/composer/metrics.py:62]: Кажется лучше это в отдельную функцию вынести (да и часть с с ts_forecasting тоже).
Возвращать из этой функции можно reference_data и results

**Review by nicl-nno — CHANGES_REQUESTED** [review]
.

File: cases/multi_target_levels_forecasting.py (hunk 1 above)
Comment (Dreamlone) [reply cases/multi_target_levels_forecasting.py]: Сделано, пример теперь оформлен через [api](https://github.com/nccr-itmo/FEDOT/pull/322/commits/6da304f3fbb5264f87cfa97cf0c9be9cdd95c8b2#diff-5f080f05d96aab5cdc73156adfcc2ca7d6cd35dbadb005a3d71e14614b02731b)

File: cases/river_levels_prediction/river_level_case_composer.py (hunk 2 above)
Comment (Dreamlone) [reply cases/river_levels_prediction/river_level_case_composer.py]: Вообще да, хотелось бы выводить то, какие параметры в узлах у операций получаются. Сделал метод "[display_structure](https://github.com/nccr-itmo/FEDOT/pull/322/commits/632f59b7832b998458b20d72e5e6e71bd99d9cd0#diff-f531f6aebd864a320d78a7369d86b8d8fd330665fced6295419574ff168664db)" у Chain

File: fedot/core/composer/metrics.py (hunk 3 above)
Comment (Dreamlone) [reply fedot/core/composer/metrics.py]: [Сделано](https://github.com/nccr-itmo/FEDOT/pull/322/commits/632f59b7832b998458b20d72e5e6e71bd99d9cd0#diff-7488db041ee4881f6441fe845b119323cb27e897f1694965bb978e1059228e66R51-R84)

File: fedot/core/chains/chain.py [hunk 4]
```diff
@@ -379,6 +379,13 @@ def _assign_data_to_nodes(self, input_data) -> Optional[InputData]:
             return None
         return input_data
 
+    def display_structure(self):
```
Comment (nicl-nno) [inline fedot/core/chains/chain.py:382]: Просто print? Чтобы было понятно, что выведется текст в консоль

File: fedot/core/composer/metrics.py [hunk 5]
```diff
@@ -42,19 +42,46 @@ class QualityMetric:
...
+
+        # Time series forecasting
+        is_ts_forecasting = reference_data.task.task_type == TaskTypesEnum.ts_forecasting
+        if is_ts_forecasting or is_multi_target_regression:
+            results, reference_data = cls.ravel_convert(results, reference_data)
+
+        return results, reference_data
+
+    @staticmethod
+    def ravel_convert(results, reference_data):
```
Comment (nicl-nno) [inline fedot/core/composer/metrics.py:69]: flatten_covert? Или у ravel есть какой-то смысл b за пределами numpy?

**Review by nicl-nno — APPROVED** [review]
Внес пару предложений по неймингу, но в целом норм.

File: fedot/core/chains/chain.py (hunk 4 above)
Comment (Dreamlone) [reply fedot/core/chains/chain.py]: [Заменил](https://github.com/nccr-itmo/FEDOT/pull/322/commits/5fe4594d68a0ec44c6e5c32304fe0ef62828ee25)

File: fedot/core/composer/metrics.py (hunk 5 above)
Comment (Dreamlone) [reply fedot/core/composer/metrics.py]: Да, вроде бы "ravel" термин довольно специфический, и не в numpy стоит избежать его использования - [исправил](https://github.com/nccr-itmo/FEDOT/pull/322/commits/5fe4594d68a0ec44c6e5c32304fe0ef62828ee25)

File: cases/multi_target_levels_forecasting.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    plt.legend(fontsize=12)
+    plt.show()
+
+    plt.plot(last_column_act, label='7th day actual')
+    plt.plot(last_column_pr, label='7th day forecast')
+    plt.ylabel('River level, cm', fontsize=14)
+    plt.xlabel('Time index', fontsize=14)
+    plt.grid()
+    plt.legend(fontsize=12)
+    plt.show()
```
Comment (J3FALL) [inline cases/multi_target_levels_forecasting.py:43]: Я бы для удобства извлек тут 2 функции

File: cases/multi_target_levels_forecasting.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    mae_value = mean_absolute_error(np.ravel(test.target), forecast)
+    print(f'MAE - {mae_value:.2f}')
+
+    if vis:
+        plot_predictions(predicted_array, test)
+
+
+if __name__ == '__main__':
+    path_file = './data/lena_levels/multi_sample.csv'
+    run_multi_output_case(path_file, vis=True)
```
Comment (J3FALL) [inline cases/multi_target_levels_forecasting.py:76]: А можно еще подобный пример, но в качестве теста?
Кажется, multi-target-регрессии у нас как раз нет в тестах

File: fedot/core/chains/chain.py [hunk 8]
```diff
@@ -379,6 +379,13 @@ def _assign_data_to_nodes(self, input_data) -> Optional[InputData]:
             return None
         return input_data
 
+    def print_structure(self):
+        """ Method print information about chain """
+        print('Chain structure:')
+        print(self.__str__())
+        for node in self.nodes:
+            print(f'{node.operation.operation_type} - {node.custom_params}')
```
Comment (J3FALL) [inline fedot/core/chains/chain.py:387]: Это сложно протестировать, но банальный вызов этой функции также можно покрыть

File: fedot/core/composer/cache.py [hunk 9]
```diff
@@ -28,10 +29,16 @@ def save_chain(self, chain):
...
+    def clear(self, tmp_only=False):
+        if not tmp_only:
+            for ext in ['bak', 'dir', 'dat']:
+                if os.path.exists(f'{self.db_path}.{ext}'):
+                    os.remove(f'{self.db_path}.{ext}')
+
+        # clear tmo files in FEDOT folder
+        temp_files = glob.glob(f'{str(default_fedot_data_dir())}/tmp_*')
+        for file in temp_files:
+            os.remove(file)
```
Comment (J3FALL) [inline fedot/core/composer/cache.py:41]: Вытащи в отдельную функцию тогда, чтобы не писать комментарий

File: fedot/core/composer/metrics.py [hunk 10]
```diff
@@ -42,19 +42,46 @@ class QualityMetric:
...
-                metric = cls.metric(reference_data, results)
+            results, reference_data = cls.prepare_data(chain, reference_data)
+            metric = cls.metric(reference_data, results)
         except Exception as ex:
             print(f'Metric evaluation error: {ex}')
         return metric
 
+    @classmethod
+    def prepare_data(cls, chain: Chain, reference_data: InputData):
+        """ Method prepare data for metric evaluation """
```
Comment (J3FALL) [inline fedot/core/composer/metrics.py:53]: кажется, тут грамматическая ошибка в описании

Comment (J3FALL) [reply fedot/core/composer/metrics.py:53]: И добавь, плз, сюда тоже тест.
Мне кажется, этот метод может еще измениться в будущем

File: fedot/core/data/data.py [hunk 11]
```diff
@@ -34,39 +34,32 @@ def from_csv(file_path=None,
...
-            pos = list(data_frame.keys()).index(target_column)
-            features = np.delete(data_array.T, [0, pos], axis=1)
+        if type(target_columns) is list:
+            features, target = process_multiple_columns(target_columns, data_frame)
         else:
-            # no target in data
-            features = data_array[1:].T
-            target = None
+            features, target = process_one_column(target_columns, data_frame,
+                                                  data_array)
```
Comment (J3FALL) [inline fedot/core/data/data.py:61]: На эти две ветки тоже можно написать соответствующий тест

**Review by J3FALL — CHANGES_REQUESTED** [review]

File: cases/multi_target_levels_forecasting.py (hunk 6 above)
Comment (Dreamlone) [reply cases/multi_target_levels_forecasting.py]: Сделал [две функции](https://github.com/nccr-itmo/FEDOT/pull/322/commits/368797c2a448de9db7670e3c2f234adc9a096960)

File: cases/multi_target_levels_forecasting.py (hunk 7 above)
Comment (Dreamlone) [reply cases/multi_target_levels_forecasting.py:103]: [Добавил ](https://github.com/nccr-itmo/FEDOT/pull/322/commits/5806d555259b3f0f7db277b3107e16409eaea392?file-filters%5B%5D=.py#diff-4308e63e89dc1627e05af132ed40cc732c63d4f331bc60fcd806be6e043d2cb4)в качестве теста в tasks/test_regression.py, но по идее его можно разместить и в composer/test_composer.py. Так весь этот фикс был как раз задуман именно потому, что композер не справлялся с задачей многомерной регрессии (вне композера у цепочек не возникало никаких проблем).

P.S. В этом [коммите](https://github.com/nccr-itmo/FEDOT/pull/322/commits/d52720cc1118f7828421ebff9a615ab84fa3f47b#diff-4308e63e89dc1627e05af132ed40cc732c63d4f331bc60fcd806be6e043d2cb4L89-L97) немного уменьшил дублирование кода

File: fedot/core/chains/chain.py (hunk 8 above)
Comment (Dreamlone) [reply fedot/core/chains/chain.py:387]: [Добавил](https://github.com/nccr-itmo/FEDOT/pull/322/commits/00120da05eb0dbfe0ab5e055443a962641accc0c#diff-e7f4408d2f7ac2deacce152f67dc72ae7a81f65d90cc53f0d0311f7b294e6c8d) соответствующий тест.

File: fedot/core/composer/cache.py (hunk 9 above)
Comment (Dreamlone) [reply fedot/core/composer/cache.py]: [Сделал](https://github.com/nccr-itmo/FEDOT/pull/322/commits/d52720cc1118f7828421ebff9a615ab84fa3f47b#diff-c50019984c360138cba52c016981875b18619ceb5defce4569bb4d33d0326e3d)

File: fedot/core/composer/metrics.py (hunk 10 above)
Comment (Dreamlone) [reply fedot/core/composer/metrics.py]: По поводу ошибки не понял (и Grammarly тоже) :) 
На всякий случай добавил "s" в prepare. Тест [добавил](https://github.com/nccr-itmo/FEDOT/pull/322/commits/d52720cc1118f7828421ebff9a615ab84fa3f47b#diff-6c1ae426c475ab8a9da8c1216d56d70a44c88ef0280208942bca070e6ca68cc0R93-R102)

File: fedot/core/data/data.py (hunk 11 above)
Comment (Dreamlone) [reply fedot/core/data/data.py:61]: [Done](https://github.com/nccr-itmo/FEDOT/pull/322/commits/d52720cc1118f7828421ebff9a615ab84fa3f47b#diff-6c1ae426c475ab8a9da8c1216d56d70a44c88ef0280208942bca070e6ca68cc0R93-R102)

File: test/unit/tasks/test_regression.py [hunk 12]
```diff
@@ -3,13 +3,24 @@
 from sklearn.metrics import mean_squared_error as mse
 
 from fedot.core.chains.chain import Chain
+from fedot.api.main import Fedot
 from fedot.core.chains.node import PrimaryNode, SecondaryNode
 from fedot.core.data.data import InputData
 from fedot.core.data.data_split import train_test_data_setup
 from fedot.core.repository.dataset_types import DataTypesEnum
 from fedot.core.repository.tasks import Task, TaskTypesEnum
```
Comment (J3FALL) [inline test/unit/tasks/test_regression.py]: Поправь импорты, есть неиспользуемые

File: fedot/core/composer/cache.py (hunk 9 above)
Comment (J3FALL) [reply fedot/core/composer/cache.py]: Я бы оформил ее в виде обычной (не static) функции, которая принимает на вход путь до folder-а и отчищает его.
А сам путь формировать в функции выше, где происходит вызов clear_folder
image

Comment (Dreamlone) [reply fedot/core/composer/cache.py]: Надеюсь правильно тебя понял - см. [тут](https://github.com/nccr-itmo/FEDOT/pull/322/commits/02f13e2da02fefb300020da140005989b378c351)

File: test/unit/tasks/test_regression.py (hunk 12 above)
Comment (Dreamlone) [reply test/unit/tasks/test_regression.py]: Если ты про строчку номер 12
image

то она используется в test_multi_target_regression_composing_correct. Не знаю, как сделать, чтобы она подсвечивалась (вроде видел подобные штуки и в других тестах у нас). Подскажи плиз, если знаешь как лучше будет сделать.

File: fedot/core/composer/cache.py (hunk 9 above)
Comment (J3FALL) [reply fedot/core/composer/cache.py]: Угу, только описание поправь
image

File: test/unit/tasks/test_regression.py (hunk 12 above)
Comment (J3FALL) [reply test/unit/tasks/test_regression.py]: Хм, да, действительно, не заметил.
Не знаю, как исправить эту штуку, мб дело в форматировании внутри Pycharm и нужно его обновить.
Но у меня та же проблема.

**Review by J3FALL — APPROVED** [review]

File: fedot/core/composer/cache.py (hunk 9 above)
Comment (nicl-nno) [reply fedot/core/composer/cache.py]: Ну тогда уж и deletes. Алсо, я бы не начинал описание с Methods - вроде и так понятно, что это он.

Comment (Dreamlone) [reply fedot/core/composer/cache.py]: [Сделано](https://github.com/nccr-itmo/FEDOT/pull/322/commits/6ed11c59f3a88efcc7189f722c8f75f34e82f7c7#diff-c50019984c360138cba52c016981875b18619ceb5defce4569bb4d33d0326e3d)
